beater/request/context.go

```diff
@@ -0,0 +1,165 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package request
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/elastic/apm-server/beater/headers"
+	logs "github.com/elastic/apm-server/log"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+const (
+	mimeTypeAny             = "*/*"
+	mimeTypeApplicationJSON = "application/json"
+)
+
+var (
+	mimeTypesJSON = []string{mimeTypeAny, mimeTypeApplicationJSON}
+)
+
+// Context abstracts request and response information for http requests
+type Context struct {
+	Req *http.Request
+
+	statusCode int
+	err        interface{}
+	stacktrace string
+
+	w http.ResponseWriter
+}
+
+// Reset allows to reuse a context by removing all request specific information
+func (c *Context) Reset(w http.ResponseWriter, r *http.Request) {
+	c.Req = r
+
+	c.w = w
+	c.statusCode = http.StatusOK
+	c.err = ""
+	c.stacktrace = ""
+}
+
+// Header returns the http.Header of the context's writer
+func (c *Context) Header() http.Header {
+	return c.w.Header()
+}
+
+// StatusCode returns the context's status code
+func (c *Context) StatusCode() int {
+	return c.statusCode
+}
+
+// Error returns the context's error
+func (c *Context) Error() interface{} {
+	return c.err
+}
+
+// Stacktrace returns the context's stacktrace
+func (c *Context) Stacktrace() string {
+	return c.stacktrace
+}
+
+// AddStacktrace sets a stacktrace for the context
+func (c *Context) AddStacktrace(stacktrace string) {
+	c.stacktrace = stacktrace
+}
+
+//TODO: All of the below methods will be changed during the response handling refactoring
```

Comment:
I didn't see where that's happening, will find you to clear it up

Reply:
That's not happening in this PR, as this PR only focuses on two things: introducing a `Context` and centralizing logging. 
As discussed I tried to keep PRs small and focused and have one PR per topic(logging, [monitoring](https://github.com/elastic/apm-server/pull/2493), [response writing](https://github.com/elastic/apm-server/pull/2494) and [package restructure](https://github.com/simitt/apm-server/pull/5)). The PRs all build on top of each other.

If you prefer I can also combine all of them into one PR, only focusing on restructuring and refactoring and not introducing logical changes (e.g. more monitoring on ACM or changing wording for logging). This would be easier to maintain for me, but result in a much bigger PR).